Support different run duration for non-timer wakeup

jhamhader opened the pull request:
# What does this implement/fix? 

In some cases, it useful to set different run durations according to
the wakeup cause, which can be found in ESP32.

For example, a sleep duration of 10 minutes to conserve power. When
waking because of timer, only run for 10 seconds. However, if waking up
due to touch or gpio, stay awake for longer (be available for action,
wait for OTA, etc.)

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** -

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** will be done later

## Test Environment

- [X] ESP32
- [X] ESP32 IDF
- [ ] ESP8266

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
deep_sleep:
  run_duration:
    default: 20s
    gpio_wakeup_reason: 10s
    touch_wakeup_reason: 15s
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [X] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs): esphome/esphome-docs#1705

oxan commented on esphome/components/deep_sleep/deep_sleep_component.cpp:
```diff
@@ -13,12 +13,34 @@ static const char *const TAG = "deep_sleep";
+}
+
 void DeepSleepComponent::setup() {
-  ESP_LOGCONFIG(TAG, "Setting up Deep Sleep...");
```
This breaks the convention of all components logging this setup message in the same style.

oxan commented on esphome/components/deep_sleep/deep_sleep_component.cpp:
```diff
@@ -13,12 +13,34 @@ static const char *const TAG = "deep_sleep";
-  if (this->run_duration_.has_value())
-    this->set_timeout(*this->run_duration_, [this]() { this->begin_sleep(); });
+  const optional<uint32_t> run_duration = get_run_duration_();
+  if (run_duration.has_value()) {
+    ESP_LOGCONFIG(TAG, "Scheduling Deep Sleep to start in %u ms", *run_duration);
+    this->set_timeout(*run_duration, [this]() { this->begin_sleep(); });
+  } else {
+    ESP_LOGCONFIG(TAG, "Deep Sleep not scheduled, run_duration is missing.");
+  }
```
So to follow the convention, I'd suggest this: (+ restore the log message above, but GitHub doesn't allow suggestions on deletions :disappointed:)
```suggestion
  if (run_duration.has_value()) {
    ESP_LOGI(TAG, "Scheduling Deep Sleep to start in %u ms", *run_duration);
    this->set_timeout(*run_duration, [this]() { this->begin_sleep(); });
  } else {
    ESP_LOGD(TAG, "Not scheduling Deep Sleep, as no run duration is configured");
  }
```
(I also lowered the priority in the `else`-branch, as it's perfectly fine to not schedule a run duration).

oxan commented on esphome/components/deep_sleep/deep_sleep_component.cpp:
```diff
@@ -29,9 +51,11 @@ void DeepSleepComponent::dump_config() {
-#ifdef USE_ESP32
-  if (wakeup_pin_ != nullptr) {
-    LOG_PIN("  Wakeup Pin: ", this->wakeup_pin_);
+#ifdef USEESP32
```
```suggestion
#ifdef USE_ESP32
  if (wakeup_pin_ != nullptr)
    LOG_PIN("  Wakeup Pin: ", this->wakeup_pin_);
```
No reason to drop the logging of the wakeup pin.

oxan reviewed (changes requested):
I like the idea, a few nits about the implementation though:

jhamhader commented:
Fixed.
Also rename 'gpio' and 'touch' to 'gpio_wakeup_reason' and 'touch_wakeup_reason'.

Will test on a device and send a documentation update soon.

jesserockz reviewed (approved):
Looks good to me, just needs docs

jhamhader committed:
Different run duration for non-timer wakeup

In some cases, it useful to set different run durations according to
the wakeup cause, which can be found in ESP32.

For example, a sleep duration of 10 minutes to conserve power. When
waking because of timer, only run for 10 seconds. However, if waking up
due to touch or gpio, stay awake for longer (be available for action,
wait for OTA, etc.)

jhamhader commented:
Tested locally, sent docs PR in esphome/esphome-docs#1705
